### ISSUE TITLE
[azservicebus] Closing a session after it has idled out can take 30 seconds.

### ISSUE THREAD

**Issue by richardpark-msft** [Azure/azure-sdk-for-go#21684]
(We noticed this while troubleshooting a separate issue)

When a link is detached by the remotes service, because it was idle, it takes up to 30 seconds for us to actually close it locally. I haven't gotten a chance to dig in for root cause. Nothing appears to be broken afterwards - the client is still usable and new senders/receivers can be created. 

Note that this doesn't happen if you force the service to detach you - for instance, updating the remote entities metadata via ARM.

**Comment by richardpark-msft** [comment]
Looking into this more the issue doesn't appear to be in go-amqp (or azservicebus). Here's some logging I added - this is basically what we see from the connection side, which eliminates a lot of code -it's just the frames being read and written on the TCP connection itself:

Our internal session object queues the session END frame

```log
2024/01/12 18:40:18 ======================= BEGIN: CLOSING Session =========================
2024/01/12 18:40:18 =====> BEGIN: s.closeOnce.Do()
2024/01/12 18:40:18 =====> session.PerformEnd BEGIN: frame sent
2024/01/12 18:40:18 =====> session.PerformEnd END: frame sent
```

In this bit, the END frame is sent on the wire. Note there is no delay here:

```log
2024/01/12 18:40:18 ====> Conn: sending frame on channel 0, frame: frames.Frame{Type:0x0, Channel:0x0, Body:(*frames.PerformEnd)(0xc000414008)}
```

After this nothing happens - there is no communication from the remote service until 30 seconds later, when we get the response to our END frame:

```log
2024/01/12 18:40:44 =====> conn: frames.PerformEnd arrived for 0
```

As far as I can tell the 30 seconds is on the service-side - we didn't receive any other frames in the meantime.

**Comment by richardpark-msft** [comment]
(NOTE for service team)


To repro this I wrote a test where I create a Sender, send one message and then wait until 10.5 minutes pass, then attempt to close the sender. I pared down my logic so that the only thing that happens is that I attempt to close the session, which initiates the END frame being sent for that AMQP session.

**Comment by richardpark-msft** [comment]
CC: @karenychen, who has been interested in progress on this since this can artificially inflate send times on idle links.

**Comment by EldertGrootenboer** [comment]
Thank you for your feedback. We have opened an investigation task for this in our backlog, and will update this issue when we have more information.

**Comment by yusundc** [comment]
My client is facing this issue and has a temporary workaround of sending/receiving a dummy message every 60 seconds. I hope this issue gets resolved soon!

**Comment by yusundc** [comment]
The issue is affecting my client as well, but they are using a temporary solution of sending and receiving a dummy message every few minutes. I look forward to a resolution soon!

**Comment by yvgopal** [comment]
@richardpark-msft "When a link is detached by the remotes service, because it was idle, it takes up to 30 seconds for us to actually close **it** locally". What is it there? I am assuming it is not the link, but the AMQP session. Is that right? I think what is meant by "it" determines what to investigate.

**Comment by richardpark-msft** [comment]
> @richardpark-msft "When a link is detached by the remotes service, because it was idle, it takes up to 30 seconds for us to actually close **it** locally". What is it there? I am assuming it is not the link, but the AMQP session. Is that right? I think what is meant by "it" determines what to investigate.

Ah, that is confusing. You are correct, it's the AMQP session. The link closes right away.

> 
> Another question. Does it happen every time a link is closed for idle time? or only sometimes?

I had a test that could cause it every time if the link idled out, but not if I forced the detach (for instance, by updating the queue using ARM).

**Comment by richardpark-msft** [comment]
We dug deeper! The main fix is in the go-amqp repo [go-amqp, PR#320](https://github.com/Azure/go-amqp/pull/320). It was a combination of a few factors: 

1. The body of the frame was smaller than 8 bytes (the session end frame is small)
2. No other packets arrive from the service.

In this case the session end body is small (7 bytes) and the service unblocked us by sending an empty frame 26 seconds later, leading to the behavior we see here.

I've got my PR merged in go-amqp, and am getting the SB PR together to incorporate it.

**Comment by richardpark-msft** [comment]
Barring any problems, release will be out tomorrow!

**Comment by richardpark-msft** [comment]
This fix has been released. There are two AMQP-based packages (azeventhubs and azservicebus):

- [azservicebus v1.6.1](https://github.com/Azure/azure-sdk-for-go/releases/tag/sdk%2Fmessaging%2Fazservicebus%2Fv1.6.1)
- [azeventhubs v1.0.4](https://github.com/Azure/azure-sdk-for-go/releases/tag/sdk%2Fmessaging%2Fazeventhubs%2Fv1.0.4)
